File: scripts/tasks.mjs
import { spawn } from "node:child_process";
import path from "node:path";
import process from "node:process";
import { fileURLToPath } from "node:url";

const root = path.dirname(path.dirname(fileURLToPath(import.meta.url)));
const node = process.execPath;

function run(command, args, options = {}) {
  return new Promise((resolve, reject) => {
    const child = spawn(command, args, {
      cwd: root,
      env: process.env,
      stdio: "inherit",
      ...options,
    });
    child.once("error", reject);
    child.once("exit", (code, signal) => {
      if (code === 0) resolve();
      else
        reject(
          new Error(
            `${path.basename(command)} failed${signal ? ` with ${signal}` : ` with exit code ${code}`}`,
          ),
        );
    });
  });
}

const runNode = (script, ...args) => run(node, [script, ...args]);
const runCargo = (...args) => run("cargo", args);

async function clean() {
  await runNode("esbuild.js", "--clean");
  await runNode("esbuild-test.js", "--clean");
}

async function build() {
  await runNode("scripts/build-browser-client.mjs");
  await runCargo("build", "--release", "-p", "fleximarkd");
  await runNode("scripts/stage-daemon.mjs");
  await runNode("scripts/create-release-manifest.mjs");
  await runNode("esbuild.js", "--production");
}

async function dev() {
  const children = [
    spawn(node, ["esbuild.js", "--watch"], {
      cwd: root,
      env: process.env,
      stdio: "inherit",
    }),
    spawn(
      node,
      [
        "node_modules/typescript/bin/tsc",
        "--noEmit",
        "--watch",
        "--project",
        "tsconfig.json",
      ],
      { cwd: root, env: process.env, stdio: "inherit" },
    ),
  ];
  let stopping = false;
  const stop = () => {
    if (stopping) return;
    stopping = true;
    for (const child of children) child.kill();
  };
  process.once("SIGINT", stop);
  process.once("SIGTERM", stop);
  await Promise.race(
    children.map(
      (child) =>
        new Promise((resolve, reject) => {
          child.once("error", reject);
          child.once("exit", (code, signal) =>
            code === 0 || stopping
              ? resolve()
              : reject(
                  new Error(
                    `development watcher failed${signal ? ` with ${signal}` : ` with exit code ${code}`}`,
                  ),
                ),
          );
        }),
    ),
  );
  stop();
}

async function compileTests() {
  await runNode("esbuild-test.js");
}

async function integrationTest() {
  await compileTests();
  await build();
  await runNode("node_modules/@vscode/test-cli/out/bin.mjs");
}

async function checkLocalization() {
  await runNode("scripts/l10n-export.mjs");
  await run("git", ["diff", "--exit-code", "--", "l10n"]);
}

async function verify() {
  await runNode("scripts/verify-architecture.mjs");
  await runNode("node_modules/typescript/bin/tsc", "--noEmit");
  await runNode(
    "node_modules/eslint/bin/eslint.js",
    "adapters",
    "web",
    "test",
    "scripts",
  );
  await checkLocalization();
  await build();
  await compileTests();
  await runNode("node_modules/@vscode/vsce/vsce", "ls", "--no-dependencies");
  await runCargo("fmt", "--all", "--", "--check");
  await runCargo("test", "--workspace", "--all-targets");
  await runCargo(
    "clippy",
    "--workspace",
    "--all-targets",
    "--",
    "-D",
    "warnings",
  );
  await runNode("scripts/check-performance-budgets.mjs");
  await runNode("node_modules/@vscode/test-cli/out/bin.mjs");
}

async function packageVsix(args) {
  await verify();
  await runNode(
    "node_modules/@vscode/vsce/vsce",
    "package",
    "--no-dependencies",
    ...args,
  );
}

const [task, ...args] = process.argv.slice(2);
const tasks = {
  build,
  clean,
  dev,
  package: () => packageVsix(args),
  test: integrationTest,
  verify,
};

if (!Object.hasOwn(tasks, task)) {
  throw new Error(`unknown task: ${task ?? "(missing)"}`);
}

await tasks[task]();

File: scripts/tasks.py
from __future__ import annotations

import argparse
import subprocess
import time
from collections.abc import Sequence
from pathlib import Path

import build as javascript_build
from _tools import ROOT, executable, run, script_entrypoint, yarn
from check_performance_budgets import check_performance_budgets
from create_release_manifest import create_manifest
from l10n_export import export_localization
from smoke_vsix import smoke_vsix
from stage_daemon import stage_daemon
from verify_architecture import verify_architecture


def clean(_: Sequence[str]) -> None:
    javascript_build.clean()


def build(_: Sequence[str]) -> None:
    javascript_build.build_browser_client()
    run("cargo", "build", "--release", "-p", "fleximarkd")
    stage_daemon()
    create_manifest()
    javascript_build.build_extension(production=True)


def stop_processes(processes: Sequence[subprocess.Popen[bytes]]) -> None:
    for process in processes:
        if process.poll() is None:
            process.terminate()
    for process in processes:
        if process.poll() is not None:
            continue
        try:
            process.wait(timeout=5)
        except subprocess.TimeoutExpired:
            process.kill()
            process.wait(timeout=5)


def dev(_: Sequence[str]) -> None:
    javascript_build.clean()
    commands = javascript_build.watch_commands()
    commands.append(
        ["yarn", "exec", "tsc", "--noEmit", "--watch", "--project", "tsconfig.json"]
    )
    processes: list[subprocess.Popen[bytes]] = []
    try:
        for command in commands:
            processes.append(
                subprocess.Popen([executable(command[0]), *command[1:]], cwd=ROOT)
            )
        while True:
            for process in processes:
                code = process.poll()
                if code is None:
                    continue
                if code != 0:
                    raise subprocess.CalledProcessError(code, process.args)
                return
            time.sleep(0.1)
    except KeyboardInterrupt:
        pass
    finally:
        stop_processes(processes)


def compile_tests() -> None:
    javascript_build.build_tests()


def integration_test(args: Sequence[str]) -> None:
    arguments = tuple(args)
    if arguments not in {(), ("--prebuilt",)}:
        raise RuntimeError("test accepts only the optional --prebuilt flag")
    compile_tests()
    if not arguments:
        build(())
    yarn("vscode-test")


def pure_test(_: Sequence[str]) -> None:
    javascript_build.build_pure_tests()
    run("node", "--test", "out/test/pure-tests.cjs")


def python_test() -> None:
    run(
        "python",
        "-m",
        "unittest",
        "discover",
        "-s",
        "scripts/tests",
        "-p",
        "test_*.py",
    )


def localization(_: Sequence[str]) -> None:
    export_localization()


def check_localization() -> None:
    localization(())
    run("git", "diff", "--exit-code", "--", "l10n")


def verify(_: Sequence[str]) -> None:
    verify_architecture()
    run("python", "-m", "compileall", "-q", "scripts")
    python_test()
    yarn("tsc", "--noEmit")
    yarn("eslint", "adapters", "web", "test", "scripts")
    check_localization()
    build(())
    compile_tests()
    run("node", "--test", "out/test/pure-tests.cjs")
    yarn("vsce", "ls", "--no-dependencies")
    run("cargo", "fmt", "--all", "--", "--check")
    run("cargo", "test", "--workspace", "--all-targets")
    run("cargo", "clippy", "--workspace", "--all-targets", "--", "-D", "warnings")
    check_performance_budgets()
    yarn("vscode-test")


def package_vsix(args: Sequence[str]) -> None:
    verify(())
    yarn("vsce", "package", "--no-dependencies", *args)


def smoke(args: Sequence[str]) -> None:
    if len(args) != 1:
        raise RuntimeError("smoke requires exactly one VSIX path")
    smoke_vsix(Path(args[0]))


TASKS = {
    "build": build,
    "clean": clean,
    "dev": dev,
    "l10n": localization,
    "package": package_vsix,
    "test-pure": pure_test,
    "smoke": smoke,
    "test": integration_test,
    "verify": verify,
}


def main() -> None:
    parser = argparse.ArgumentParser(description="Run FlexiMark development tasks")
    parser.add_argument("task", choices=TASKS)
    parser.add_argument("task_args", nargs=argparse.REMAINDER)
    args = parser.parse_args()
    TASKS[args.task](args.task_args)


if __name__ == "__main__":
    script_entrypoint(main)
